File: care_connect_flutter_frontend/lib/theme.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';

// ── Design tokens ──────────────────────────────────────────────────────────────

class CTokens {
  // Shared brand
  static const primary = Color(0xFF357C6F);
  static const primaryLight = Color(0xFF2F7A6B);
  static const caregiverPurple = Color(0xFF684BE6);
  static const danger = Color(0xFFC53030);
  static const planIndigo = Color(0xFF6366F1);

  // Dark palette (matches Figma dark imports)
  static const darkBg       = Color(0xFF0E131D);
  static const darkSurface  = Color(0xFF1D2534);
  static const darkSurface2 = Color(0xFF141A27);
  static const darkBorder   = Color(0xFF333D4F);
  // Boundary of an interactive control (button outline, chip, switch track,
  // input). Clears 3:1 against darkBg, darkSurface and darkSurface2, which
  // darkBorder (1.70:1 on darkBg) never did — WCAG 2.1 SC 1.4.11.
  static const darkControlBorder = Color(0xFF5D6F90);
  static const darkInputBorder = Color(0xFF7A8599);
  static const darkText     = Color(0xFFF5F7FA);
  static const darkSub      = Color(0xFF9EA8BD);
  static const darkMuted    = Color(0xFF828DA1);
  static const darkLink     = Color(0xFF59AD9E);

  // Light palette (matches Figma light imports)
  static const lightBg       = Color(0xFFF6F7F9);
  static const lightSurface  = Color(0xFFF2F3F5);
  static const lightSurface2 = Color(0xFFE8EAED);
  static const lightBorder   = Color(0xFFE0E2E6);
  // See darkControlBorder. The old value was 1.17:1 against lightSurface,
  // leaving inputs and outlined buttons with no detectable boundary.
  static const lightControlBorder = Color(0xFF7E8697);
  static const lightInputBorder = Color(0xFF7E8697);
  static const lightText     = Color(0xFF1A2133);
  static const lightSub      = Color(0xFF5A6478);
  // Nudged one step darker: the old #666E7D was 4.26:1 on lightSurface2,
  // just under the 4.5:1 floor for the message hint and disabled button text.
  static const lightMuted    = Color(0xFF626A78);
  static const lightLink     = Color(0xFF2F7A6B);
}

//this enables the changing of the theme between light and dark mode
class CScheme {
  final Color bg, surface, surface2, border, inputBorder;
  final Color text, sub, muted, link, primary;

  /// Boundary colour for interactive controls, guaranteed to clear 3:1 against
  /// every surface in this scheme (SC 1.4.11). [border] stays the lighter,
  /// decorative outline used for cards and dividers, which the success
  /// criterion does not cover.
  final Color controlBorder;

  const CScheme({
    required this.bg, required this.surface, required this.surface2,
    required this.border, required this.inputBorder, required this.text,
    required this.sub, required this.muted, required this.link, required this.primary,
    required this.controlBorder,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CScheme &&
          bg == other.bg && surface == other.surface && surface2 == other.surface2 &&
          border == other.border && inputBorder == other.inputBorder &&
          text == other.text && sub == other.sub && muted == other.muted &&
          link == other.link && primary == other.primary &&
          controlBorder == other.controlBorder;

  @override
  int get hashCode => Object.hash(bg, surface, surface2, border, inputBorder, text, sub, muted, link,
      primary, controlBorder);

  static const dark = CScheme(
    bg: CTokens.darkBg, surface: CTokens.darkSurface, surface2: CTokens.darkSurface2,
    border: CTokens.darkBorder, inputBorder: CTokens.darkInputBorder,
    controlBorder: CTokens.darkControlBorder,
    text: CTokens.darkText, sub: CTokens.darkSub, muted: CTokens.darkMuted,
    link: CTokens.darkLink, primary: CTokens.primary,
  );

  static const light = CScheme(
    bg: CTokens.lightBg, surface: CTokens.lightSurface, surface2: CTokens.lightSurface2,
    border: CTokens.lightBorder, inputBorder: CTokens.lightInputBorder,
    controlBorder: CTokens.lightControlBorder,
    text: CTokens.lightText, sub: CTokens.lightSub, muted: CTokens.lightMuted,
    link: CTokens.lightLink, primary: CTokens.primaryLight,
  );
}

// ── Contrast helpers (WCAG 2.1 SC 1.4.3 / 1.4.11) ─────────────────────
//
// The design leans on a "tinted pill" pattern everywhere: a chip, badge or
// secondary button painted with `accent.withValues(alpha: 0.12)` and then
// labelled in that same `accent`. That is text on a tint of itself, and it
// measured as low as 1.77:1 against the required 4.5:1.
//
// Rather than hand-pick a second colour for every accent in both themes -
// and get it wrong again the next time someone adds one - `readableOn`
// derives the label colour at build time. Appointment type colours are
// arbitrary, so a runtime rule is the only thing that covers them all.

/// Contrast ratio between two opaque colours, per the WCAG 2.1 definition.
double contrastRatio(Color a, Color b) {
  final la = a.computeLuminance();
  final lb = b.computeLuminance();
  final hi = math.max(la, lb);
  final lo = math.min(la, lb);
  return (hi + 0.05) / (lo + 0.05);
}

/// Composites a translucent [fg] over an opaque [bg], giving the colour a
/// reader actually sees. Needed because the tint backgrounds are semi
/// transparent, and contrast must be measured against the composite.
Color compositeOver(Color fg, Color bg) => Color.alphaBlend(fg, bg);

/// Returns [base] shifted in lightness — hue and saturation untouched — until
/// it reaches [minRatio] against [background].
///
/// Darkens when the background is light and lightens when it is dark, and
/// returns [base] unchanged when it already passes, so colours that were
/// already compliant keep their exact designed value.
///
/// [minRatio] defaults to the 4.5:1 floor for body text; pass 3.0 for large
/// text (SC 1.4.3) or for non-text boundaries and indicators (SC 1.4.11).
Color readableOn(Color base, Color background, {double minRatio = 4.5}) {
  if (contrastRatio(base, background) >= minRatio) return base;

  final hsl = HSLColor.fromColor(base);
  final lighten = background.computeLuminance() <= 0.18;
  var best = base;

  for (var i = 1; i <= 100; i++) {
    final t = i / 100;
    final l = lighten
        ? hsl.lightness + (1 - hsl.lightness) * t
        : hsl.lightness * (1 - t);
    final candidate = hsl.withLightness(l.clamp(0.0, 1.0)).toColor();
    best = candidate;
    if (contrastRatio(candidate, background) >= minRatio) return candidate;
  }
  return best;
}

/// Chooses a legible label colour for a solid [fill], and nudges the fill
/// itself only when neither white nor near-black can reach [minRatio] on it.
///
/// Needed for avatar badges, whose fill is a person's identity colour chosen
/// for recognition rather than contrast: white initials on the amber contact
/// measured 2.15:1. Flipping the label to near-black fixes that one outright
/// and keeps the identity colour intact; only a fill that defeats both label
/// colours (mid-tone indigo, 4.47:1) gets adjusted.
({Color fill, Color foreground}) legibleOn(Color fill, {double minRatio = 4.5}) {
  const dark = Color(0xFF1A2133);
  final useWhite =
      contrastRatio(const Color(0xFFFFFFFF), fill) >= contrastRatio(dark, fill);
  final foreground = useWhite ? const Color(0xFFFFFFFF) : dark;
  return (
    fill: readableOn(fill, foreground, minRatio: minRatio),
    foreground: foreground,
  );
}

/// Convenience for the tinted-pill pattern: given the [accent] a pill is
/// tinted with, the [surface] it sits on and the tint [alpha], returns the
/// label colour that clears [minRatio] against the composited pill.
Color readableOnTint(Color accent, Color surface, double alpha,
        {double minRatio = 4.5}) =>
    readableOn(accent, compositeOver(accent.withValues(alpha: alpha), surface),
        minRatio: minRatio);

// ── Theme notifier ──────────────────────────────────────────────────────────────

class ThemeNotifier extends ChangeNotifier {
  bool _dark;

  ThemeNotifier({bool isDark = true}) : _dark = isDark;

  bool get isDark => _dark;
  CScheme get scheme => _dark ? CScheme.dark : CScheme.light;

  void toggle() { _dark = !_dark; notifyListeners(); }
  void setDark(bool value) { _dark = value; notifyListeners(); }
}

// ── Material ThemeData helpers ──────────────────────────────────────────────────

ThemeData buildTheme(bool isDark) {
  final s = isDark ? CScheme.dark : CScheme.light;
  final base = isDark ? ThemeData.dark() : ThemeData.light();
  return base.copyWith(
    scaffoldBackgroundColor: s.bg,
    colorScheme: ColorScheme(
      brightness: isDark ? Brightness.dark : Brightness.light,
      primary: s.primary,
      onPrimary: Colors.white,
      secondary: CTokens.planIndigo,
      onSecondary: Colors.white,
      surface: s.surface,
      onSurface: s.text,
      error: CTokens.danger,
      onError: Colors.white,
    ),
    textTheme: GoogleFonts.interTextTheme(base.textTheme).apply(
      bodyColor: s.text,
      displayColor: s.text,
    ),
    dividerColor: s.border,
    cardColor: s.surface,
  );
}

File: CareConnectReact/src/constants/theme.ts
// CareConnect design tokens — mirrors Flutter's CTokens / CScheme
export const tokens = {
  primary:          '#357C6F',
  primaryLight:     '#2F7A6B',
  caregiverPurple:  '#684BE6',
  danger:           '#C53030',
  planIndigo:       '#6366F1',
  white:            '#FFFFFF',
} as const;

export interface ColorScheme {
  bg:           string;
  surface:      string;
  surface2:     string;
  border:       string;
  inputBorder:  string;
  text:         string;
  sub:          string;
  muted:        string;
  link:         string;
  primary:      string;
  /** Use for text/icon glyphs colored with the brand hue (WCAG AA 4.5:1 on
   *  bg/surface/surface2). `primary` itself stays reserved for solid fills
   *  and decorative borders/icons, where only the 3:1 non-text threshold
   *  applies and lightening it would weaken white-on-primary button text. */
  primaryText:  string;
  /** Use for danger-colored text/icon glyphs (WCAG AA 4.5:1). `tokens.danger`
   *  stays reserved for solid fills (e.g. destructive button backgrounds). */
  danger:       string;
}

export const dark: ColorScheme = {
  bg:          '#0E131D',
  surface:     '#1D2534',
  surface2:    '#141A27',
  border:      '#333D4F',
  inputBorder: '#7A8599',
  text:        '#F5F7FA',
  sub:         '#9EA8BD',
  muted:       '#828DA1',
  link:        '#59AD9E',
  primary:     '#357C6F',
  primaryText: '#5D968C',
  danger:      '#D66C6C',
};

export const light: ColorScheme = {
  bg:          '#F6F7F9',
  surface:     '#F2F3F5',
  surface2:    '#E8EAED',
  border:      '#E0E2E6',
  inputBorder: '#E0E2E6',
  text:        '#1A2133',
  sub:         '#5A6478',
  muted:       '#626A78',
  link:        '#2D7567',
  primary:     '#2D7567',
  primaryText: '#2D7567',
  danger:      '#C53030',
};

// ── Accessible accent text color ────────────────────────────────────────────
// Category/severity chips borrow one-off accent hues (e.g. appointment type
// colors, symptom severity colors) that were only tuned for use as tinted
// fills/borders, not as text. When one of those hues is drawn as *text*,
// run it through this helper first so it meets WCAG AA 4.5:1 against every
// background in the active scheme, instead of hand-tuning each hue per theme.

function hexToRgb(hex: string): [number, number, number] {
  const h = hex.replace('#', '');
  return [parseInt(h.slice(0, 2), 16), parseInt(h.slice(2, 4), 16), parseInt(h.slice(4, 6), 16)];
}

function rgbToHex([r, g, b]: [number, number, number]): string {
  const c = (v: number) => Math.max(0, Math.min(255, Math.round(v))).toString(16).padStart(2, '0');
  return `#${c(r)}${c(g)}${c(b)}`.toUpperCase();
}

function channelLuminance(c: number): number {
  const v = c / 255;
  return v <= 0.03928 ? v / 12.92 : ((v + 0.055) / 1.055) ** 2.4;
}

function relativeLuminance([r, g, b]: [number, number, number]): number {
  return 0.2126 * channelLuminance(r) + 0.7152 * channelLuminance(g) + 0.0722 * channelLuminance(b);
}

function contrastRatio(hexA: string, hexB: string): number {
  const lA = relativeLuminance(hexToRgb(hexA));
  const lB = relativeLuminance(hexToRgb(hexB));
  const [lighter, darker] = lA > lB ? [lA, lB] : [lB, lA];
  return (lighter + 0.05) / (darker + 0.05);
}

/**
 * Returns `hex` unchanged if it already reads at 4.5:1 against every
 * background in `scheme`; otherwise mixes it toward white (dark schemes) or
 * black (light schemes) until it does, so accent hues stay legible as text
 * in both themes without needing a hand-picked value per hue per theme.
 */
export function accessibleAccentText(hex: string, scheme: ColorScheme): string {
  const backgrounds = [scheme.bg, scheme.surface, scheme.surface2];
  const isDarkScheme = relativeLuminance(hexToRgb(scheme.bg)) < 0.5;
  const mixTarget: [number, number, number] = isDarkScheme ? [255, 255, 255] : [0, 0, 0];

  const base = hexToRgb(hex);
  const worstContrast = () => Math.min(...backgrounds.map(bg => contrastRatio(rgbToHex(mixed), bg)));

  let mixed = base;
  if (Math.min(...backgrounds.map(bg => contrastRatio(hex, bg))) >= 4.5) return hex;

  for (let i = 1; i <= 100; i++) {
    const t = i / 100;
    mixed = [0, 1, 2].map(ch => base[ch] + (mixTarget[ch] - base[ch]) * t) as [number, number, number];
    if (worstContrast() >= 4.5) break;
  }
  return rgbToHex(mixed);
}

/**
 * Returns `hex` unchanged if white text already reads at 4.5:1 on top of it,
 * otherwise darkens it just enough that white text does. Use this for a
 * solid accent-colored fill (e.g. a selected severity dot) that carries
 * white text/icon content, instead of hand-picking a darker shade per hue.
 */
export function accessibleFillForWhiteText(hex: string): string {
  if (contrastRatio('#FFFFFF', hex) >= 4.5) return hex;
  const base = hexToRgb(hex);
  let mixed = base;
  for (let i = 1; i <= 100; i++) {
    const t = i / 100;
    mixed = [0, 1, 2].map(ch => base[ch] * (1 - t)) as [number, number, number];
    if (contrastRatio('#FFFFFF', rgbToHex(mixed)) >= 4.5) break;
  }
  return rgbToHex(mixed);
}
